lib/Sema/CSDiag.cpp

```diff
@@ -5372,6 +5382,137 @@ bool FailureDiagnosis::diagnoseCallContextualConversionErrors(
   return false;
 }
 
+// Offer a fix-it for using `subscript` instead of the operator
+// if the inputs are compatible.
+static bool diagnoseSubscriptMisuse(ApplyExpr *callExpr,
+                                    ConstraintSystem &CS,
+                                    FailureDiagnosis &FD) {
+  auto UDE = dyn_cast<UnresolvedDotExpr>(callExpr->getFn());
+  if (!UDE)
+    return false;
+
+  auto baseExpr = UDE->getBase();
+  if (baseExpr && UDE->getName().getBaseName() == "subscript") {
+    auto baseType = CS.getType(baseExpr);
+    // Look up subscript declarations and return if there are none.
+    auto lookup = CS.lookupMember(baseType->getRValueType(),
+                                  DeclName(DeclBaseName::createSubscript()));
+    if (lookup.empty())
+      return false;
+
+    auto argExpr = FD.typeCheckChildIndependently(callExpr->getArg(),
```

Comment:
Since you are making a use of both FD and CS it might make sense to make this a method on `FailureDiagnosis` so both are easily accessible.

Reply:
Why not. But then I wonder why `typeCheckChildIndependently` is public.

Reply:
As you can see access control on these methods hasn't really been thought through.